force-app/main/default/classes/P1.cls

```diff
@@ -1,45 +1,141 @@
-        return 0;
-    }
-    
-    /**
-     * Given a string, return its backwards order.
-     * */
-    public static String reverseString(String s) {
-        return '';
-    }
-    
-    /**
-     * Use recursion to compute the factorial of a given number.
-     * */
-    public static Integer factorial(Integer n) {
-        return 0;
-    }
-    
-    /**
-     * Given a string, determine if it is a palindrome.
-     * */
-    public static Boolean isPalindrome(String s) {
-        return false;
-    }
-    
-    /**
-     *  Given a state, return the accounts in that state.
-     *  I only want the accounts that have a contact who works at the same place as the account.
-     * */
-    public static List<Account> getAccountsForAGivenState(State s) {
-        return new List<Account>();
-    }
-    
-    /**
-     * For each student, get the grade for each of their classes.
-     * */
-    public static Map<Id, Map<Id, Double>> getClassGrades() {
-        return new Map<Id, Map<Id, Double>>();
-    }
+	/**
+	 * Return the nth fiboacci number in the fibonacci sequence.
+	 * It should start 1,1,2,3...
+	 * */
+	public static Integer fibonacci(Integer n) {
+		if(n == 0 || n == 1) {
+			return n;
+		}
+		return fibonacci(n - 1) + fibonacci(n - 2);
+	}
+
+	/**
+	 * Given a string, return its backwards order.
+	 * */
+	public static String reverseString(String s) {
+		String revStr = '';
+		for(Integer i =(s.length() - 1); i >= 0; i --) {
+			revStr += s.substring(i, i + 1);
+		}
+		return revStr;
+	}
+
+	/**
+	 * Use recursion to compute the factorial of a given number.
+	 * */
+	public static Integer factorial(Integer n) {
+		if(n < 0) {
+			return null;
+		}
+
+		if(n == 0) {
+			return 1;
+		}
+		return n * factorial(n - 1);
+	}
+
+	/**
+	 * Given a string, determine if it is a palindrome.
+	 * */
+	public static Boolean isPalindrome(String s) {
+		if(s.equals('')) {
+			return true;
+		}
+		Integer forw = 0;
+		Integer backw = s.length() - 1;
+		// remove punctuation, convert to lowercase
+		Pattern nonAlphanumeric = Pattern.compile('[^a-zA-Z0-9]');
+		Matcher matcher = nonAlphanumeric.matcher(s);
+		s = matcher.replaceAll('').toLowerCase();
+		while(backw > forw && backw < s.length() && forw < s.length()) {
+			String fChar = s.substring(forw, forw + 1);
+			String bChar = s.substring(backw, backw + 1);
+			if(!fChar.equals(bChar)) {
+				return false;
+			}
+			forw++;
+			backw--;
+		}
+		return true;
+	}
+
+	/**
+	 *  Given a state, return the accounts in that state.
+	 *  I only want the accounts that have a contact who works at the same place as the account.
+	 * */
+	public static List<Account> getAccountsForAGivenState(State s) {
+		// select accounts for whose billig state is s
+		List<Account> accounts = [SELECT Id, Name, BillingState FROM Account WHERE BillingState = :s.name()];
+		// select all accounts
+		List<Contact> contacts = [SELECT Id, Name, AccountId, MailingState FROM Contact];
```

Comment:
You're grabbing ALL of the Contacts inside of the Org. Although it may not be an issue now, when you have thousands/millions of records, there will be a lost of wasted resources.

Reply:
You should instead collect the Ids inside of List<Id> and use that list to only query for the Contacts whose accounts are inside of that list.

`
List<Id> accountIds = new List<Id>();
for(Account a: accounts) {
    accountIds.add(a.Id);
}

List<Contact> contacts = [SELECT Id, Name, AccountId, MailingState FROM Contact WHERE Id IN: accountIds];
`
Now, the contacts list only grabs the Contacts that are relevant.